### PR TITLE
Tutorials infrastructure: switch to text-based notebook workflow

### PR DESCRIPTION
**Pull request by alexdesiqueira**
Changes the workflow to use a text-based format for Jupyter notebooks and sphinx+myst for running and testing them, continuing changes added by @jni.
Kindly stolen from [@rossbar's numpy-tutorials](https://github.com/numpy/numpy-tutorials/pull/36/).

### CODE REVIEW

**Comment by alexdesiqueira** [comment]
Website looks like this so far:
image

**Comment by alexdesiqueira** [comment]
How it looks so far:
image

**Comment by alexdesiqueira** [comment]
Alright, I think that's it as a starting point.
We need to discuss how the tutorials will be structured, but the main machinery is here already.

File: _toc.yml [hunk 1]
```diff
@@ -1,6 +1,7 @@
-- file: lectures/00_images_are_arrays
+format: jb-article
```
Comment (jni) [inline _toc.yml:1]: Can we use `jb-book` instead?

File: content/00_images_are_arrays.md [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+|2D grayscale|(row, column)|
+|2D multichannel|(row, column, channel)|
+|3D grayscale (or volumetric) |(plane, row, column)|
+|3D multichannel|(plane, row, column, channel)|
+
++++
+
+## Displaying images using matplotlib
+
+```{code-cell} ipython3
```
Comment (jni) [inline content/00_images_are_arrays.md:87]: Can these just say "python"?

**Review by jni — COMMENTED** [review]
@alexdesiqueira in addition to the minor comments below, why are the existing tutorials not migrated here? I would prefer that ongoing PRs only add content rather than removing it! 😂

**Comment by alexdesiqueira** [comment]
> why are the existing tutorials not migrated here? I would prefer that ongoing PRs only add content rather than removing it!

We're not removing anything yet! :joy: If I convert _everything_ now, this PR will be huge! The idea here was only to add the infrastructure — the tutorial(s) presented here are examples :slightly_smiling_face:

File: _toc.yml (hunk 1 above)
Comment (alexdesiqueira) [reply _toc.yml:1]: Sure!

File: content/00_images_are_arrays.md (hunk 2 above)
Comment (alexdesiqueira) [reply content/00_images_are_arrays.md:87]: I'll test it; I converted these automatically.

File: .github/workflows/deploy.yml [hunk 3]
```diff
@@ -60,17 +61,17 @@ jobs:
...
 
-      # Deploy the book's HTML to github pages
+      # Deploy site to github pages
       - name: GitHub Pages action
         if: github.repository == 'scikit-image/skimage-tutorials' && github.ref == 'refs/heads/main' && github.event_name == 'push'
         uses: peaceiris/actions-gh-pages@v3.6.1
         with:
           github_token: ${{ secrets.GITHUB_TOKEN }}
-          publish_dir: ./_build/html
+          publish_dir: ./website/_build/html
```
[the remark this answers is not in the document]
Comment (alexdesiqueira) [reply .github/workflows/deploy.yml:76]: I felt it'd make things more organized. I don't mind, though. Would you like me to leave it as it was?

**Comment by jni** [comment]
I didn't want you to convert everything, just the things that I already converted to the book! 😜

**Comment by alexdesiqueira** [comment]
@jni ah, of course :laughing: Forgot there are some around! I'll put them here, too

**Comment by alexdesiqueira** [comment]
@jni just a FYI: the notebooks at #53 were not converted to MyST. I don't recall exactly what we did at that day, but I added more notebooks: the ones at `not_yet_booked`. Let me know if there's something I didn't understand yet — as in, are there notebooks converted to MyST somewhere else that I should look for :slightly_smiling_face:

**Comment by alexdesiqueira** [comment]
Also FYI, Alex-from-the-future: to convert from `ipynb` to MyST, the command is:
```bash
jupytext <notebook_to_be_converted.ipynb> --to myst
```
You can also `*` a large folder full of notebooks :slightly_smiling_face:

**Comment by alexdesiqueira** [comment]
Starting to smell that moldy odor :slightly_smiling_face: @jni @stefanv any consensus on this one?

**Review by jni — APPROVED** [review]
I'm very happy with this, sorry that I forgot about it! 😂

**Comment by jni** [comment]
@alexdesiqueira can you push an empty commit to retry CI? I can't seem to be able to manually rebuild...

**Comment by alexdesiqueira** [comment]
At "Build the book," I'm receiving:

```
Run jupyter-book build .
Running Jupyter-Book v0.12.1
(...)
File "/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/jupyter_book/utils.py", line 48, in _error
    raise kind(box)
RuntimeError: 
===============================================================================

The Table of Contents file is malformed: toc is not a mapping: <class 'list'>
You may need to migrate from the old format, using:

	jupyter-book toc migrate /home/runner/work/skimage-tutorials/skimage-tutorials/_toc.yml -o /home/runner/work/skimage-tutorials/skimage-tutorials/_toc.yml

===============================================================================
```

Maybe something we changed in #53?

**Comment by alexdesiqueira** [comment]
Thank you @stefanv!

**Comment by stefanv** [comment]
Something isn't right with the caching: it rebuilds all notebooks on each make.

**Comment by stefanv** [comment]
The Makefile is definitely not right... refers to a `lessons` dir that does not exist, directories are off, etc.

**Comment by alexdesiqueira** [comment]
> The Makefile is definitely not right... refers to a `lessons` dir that does not exist, directories are off, etc.

@stefanv thanks; gonna check that.

**Comment by stefanv** [comment]
I will push up a patch, then you can work on top of that.

**Comment by stefanv** [comment]
Actually, the whole thing is completely broken :)

**Comment by alexdesiqueira** [comment]
:joy: Sorry about that.

**Comment by stefanv** [comment]
OK, I pushed up a new Makefile and switched to using jupytext to render directly.  Jupytext correctly passes through failing cells with the `raises-exception` tag. Other tags (`remove-input`, `remove-output`) are handled by `myst_nb`.

Process now is:
1. Files are stored as myst notebooks (.md)
2. Jupytext converts to ipynb and executes inline
3. myst_nb renders those notebooks inside Sphinx

There's no caching, but make will not rebuild a notebook unless it has changed.  I.e., if a .md source file is modified, the notebook is re-rendered and re-executed.

**There will be failures on CI**.  These are real failures that need to be fixed.

**Comment by rossbar** [comment]
Adding my two cents from afar...

Though the build system is a departure from the jupyter-book or pure-sphinx workflows, it LGTM and shouldn't make a difference to tutorial authors one way or the other once everything is working. From trying it out, there are two main features that you get from the sphinx/jb workflows that are missing here:
 - For whatever reason, the execution caching doesn't seem to be working. It seems to me based on the Makefile that the notebooks should only be executed on a re-build if their content has changed, but in practice all the notebooks are re-executed on each build. Maybe the `myst-nb` sphinx extension is `touch`-ing or otherwise modifying the timestamps of the generated `.ipynb` files during the sphinx build? It's not immediately obvious to me why this isn't working...
 - Another feature for the build system that would be nice would be the ability to continue building even after execution of a notebook fails. Currently if the `jupytext --execute` fails, then the full `make` command exits and the outputs aren't generated. With the sphinx workflow, the default behavior is that execution failures are treated as warnings and the full execute-and-convert-to-html procedure continues to completion. This is a nice feature because authors/reviewers can see the execution failures in the output html, which IME has helped a lot with reviewing etc. in `numpy-tutorials` and `nx-guides`.

**Comment by stefanv** [comment]
@rossbar do you want to update the build system to use the Sphinx workflow? Fine with me, as long as we have the ability to error on failed build for CI purposes.

**Comment by stefanv** [comment]
I am a bit surprised at the rebuilds. That's the whole point of make, but I observe the same thing.

**Comment by rossbar** [comment]
> do you want to update the build system to use the Sphinx workflow? 

Sure, I'm happy to do so, if for no other reason than to provide a point of comparison.

> Fine with me, as long as we have the ability to error on failed build for CI purposes.

Yeah this is a bit of a sore point: I do this with the `SPHINXOPTS=-W --keep-going` flags: -W to elevate warnings to errors, and `--keep-going` to ensure that everything runs to completion. The downside is that it elevates *all* sphinx warnings to errors, not just execution failures. FWIW this is a requested enhancement in myst-nb, see e.g. executablebooks/myst-nb#248

**Comment by alexdesiqueira** [comment]
Superseded by #66; closing this one in favor of #66 then. Thank you @rossbar!

**Pull request closed without merging**